app/assets/javascripts/browsered/web-payments/google-pay.js

```diff
@@ -39,6 +44,95 @@ const processPayment = paymentData => {
     })
 }
 
+const checkOriginHostName = origin => {
+  return Charge.worldpay_3ds_flex_ddc_url.lastIndexOf(origin, 0) === 0
+}
+
+const performDeviceDataCollectionForGooglePay = (paymentData) => {
+  const deviceDataCollectionFinishedEventListener = event => {
+    if (checkOriginHostName(event.origin) && event.source === iframeWindow) {
+      const { MessageType, SessionId, Status } = JSON.parse(event.data)
+      if (MessageType === 'profile.completed') {
+        var ddcStatus = null
+        var ddcResult = null
+        if (Status === true && typeof SessionId === 'string') {
+          ddcStatus = 'valid DDC result'
+          ddcResult = SessionId
+        } else if (!Status) {
+          ddcStatus = 'DDC result did not have Status of true'
+        } else {
+          ddcStatus = 'no SessionID string in DDC result'
+        }
+        Charge.googlePayWorldpay3dsFlexDeviceDataCollectionResult = ddcResult
+        window.clearTimeout(deviceDataCollectionTimeout)
+        window.removeEventListener('message', deviceDataCollectionFinishedEventListener)
+        submitGooglePayAuthRequest(paymentData, ddcStatus, ddcResult)
+      }
+    }
+  }
+
+  const DDC_TIMEOUT_IN_MILLISECONDS = 30000
+
+  const deviceDataCollectionTimeout = window.setTimeout(() => {
+    window.removeEventListener('message', deviceDataCollectionFinishedEventListener)
+    submitGooglePayAuthRequest(paymentData, `DDC timeout after ${DDC_TIMEOUT_IN_MILLISECONDS} milliseconds`)
+  }, DDC_TIMEOUT_IN_MILLISECONDS)
+
+  var iframe = document.getElementById('googlePayWorldpay3dsFlexDdcIframe')
+
+  try {
+    // eslint-disable-next-line no-unused-expressions
+    iframe.contentWindow.location.href
```

Comment:
What is this expected to do?

Reply:
Same-origin locks access to href. When trying to access href and it throws an exception, then we know that we have to re-create the iframe as it has been changed by Google.

Reply:
Do you mean changed by DDC? Maybe add a comment to explain it?

Reply:
👍